pkgs/development/python-modules/exif/default.nix

```diff
@@ -0,0 +1,35 @@
+{ lib, buildPythonPackage, fetchPypi, isPy3k, plum_py }:
+
+let
+  # exif depends on older plum_py==0.3.1
+  plum_py' = plum_py.overridePythonAttrs (old: rec {
+    version = "0.3.1";
+    src = fetchPypi {
+      inherit (old) pname format;
+      inherit version;
+      python = "py3";
+      sha256 = "1ra8b9rqh3ykbryqn9np03i0d42m8xcxh13by72k8q8409syqa4k";
+    };
+  });
+in
```

Comment:
Pinning within python-modules is highly discouraged. This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

The preference to handling this is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

Reply:
Thanks for the suggestions. The main goal is an application, gphotos-sync, and the packages that I'm adding don't currently have other users in nixpkgs, but they're not specific to that application, so I think the overlay approach would be inappropriate here.

I've redone things using fetchFromGitLab/GitHub to get the full source for tests, and relaxed the version requirement so that everything is using the latest versions.